File: leetcode/p00050/p00050.go
package p00050

import "math"

// Time Complexity: O(logn)
// Space Complexity: O(1)
func myPow(x float64, n int) float64 {
	if x == 0 {
		return 0
	}
	if n == 0 {
		return 1
	}

	neg := false
	if n < 0 {
		n = -n
		neg = true
	}

	var ans float64 = 1
	for n != 0 {
		if n%2 != 0 {
			ans *= x
		}
		x *= x
		n >>= 1
	}

	if neg {
		ans = 1 / ans
	}
	return math.Round(ans*100000) / 100000
}

File: leetcode/p00050/p00050.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 0:
            return 0
        if n == 0:
            return 1
        
        neg = False
        if n < 0:
            n = -n
            neg = True
        
        ans = 1
        while n != 0:
            if n%2 == 1:
                ans *= x
            x *= x
            n >>= 1
        
        if neg:
            ans = 1/ans
        return round(ans*100000)/100000

if __name__ == '__main__':
    s = Solution()

    assert s.myPow(2.00000, 4) == 16.00000
    assert s.myPow(2.00000, 10) == 1024.00000
    assert s.myPow(2.10000, 3) == 9.26100